lib/tensorflow/session.rb

```diff
@@ -23,96 +23,26 @@ def initialize()
   end
 
   #
-  # Converts a give ruby array to C array (using SWIG) by detecting the data type automatically. 
-  # Design decision needs to be made regarding this so the all the data types are supported.
-  # Currently Integer(Ruby) is converted to long long(C) and Float(Ruby) is converted double(C).
-  #
-  # * *Returns* :
-  #   - A c array.
-  #
-  def graph_def_to_c_array(array)
-   c_array = Tensorflow::Character.new(array.length)
-   (0..array.length-1).each do |i|
-     c_array[i] = array[i]
-   end
-   c_array
-  end
-
-  #
   # Runs a session on a given input.
   # * *Returns* :
   #   - nil
   #
   def run(inputs, outputs, targets)
-  	inputNames = Tensorflow::String_Vector.new
-  	inputValues = Tensorflow::Tensor_Vector.new
-    if inputs != nil
-  	  inputs.each do |key, value|
-  		  inputValues.push(value)
-  		  inputNames.push(key)
-  	  end
-    end
-
-  	outputNames = Tensorflow::String_Vector.new
-  	outputs.each do |name|
-  		outputNames.push(name)
-  	end
+    input_names, input_values = initialize_inputs(inputs)
+    output_names, output_values = initialize_outputs(outputs)
+    target_names = initialize_targets(targets)
 
-  	targetNames = Tensorflow::String_Vector.new
-    if targets != nil
-  	  targets.each do |name|
-  		  targetNames.push(name)
-  	  end
-    end
-
-  	outputValues = Tensorflow::Tensor_Vector.new
    	status = Tensorflow::TF_NewStatus()
-	  Tensorflow::TF_Run_wrapper(self.session , inputNames, inputValues, outputNames, outputValues, targetNames, self.status)
+	  Tensorflow::TF_Run_wrapper(self.session, input_names, input_values, output_names, output_values, target_names, self.status)
```

Comment:
(Sorry, I know the PR is already merged, and it is very clean and reasonable, but I'd like to discuss further possible improvements.)

1. `self` is unnecessary here.
2. For me, it seems like internal object, being an "implementation", should not be called the same way as the current class, it may lead to confusion ("this is session, but not of class Session?.. hmmmm"). I prefer to call such an object `internal` or `implementation`.

WDYT of this line looking like (note the additional namespace too):
```ruby
Tensorflow::Internal::TF_Run_wrapper(internal, input_names, input_values, output_names, output_values, target_names, status)
```